modules/rostests/kmtests/hal/HalPortIo.c

```diff
@@ -0,0 +1,193 @@
+/*
+ * PROJECT:     ReactOS Kernel-Mode Tests
+ * LICENSE:     LGPL-2.1-or-later (https://spdx.org/licenses/LGPL-2.1-or-later)
+ * PURPOSE:     Tests for string I/O intrinsic functions
+ * COPYRIGHT:   Copyright 2025 Dmitry Borisov <di.sean@protonmail.com>
+ */
+
+/* INCLUDES *******************************************************************/
+
+#include <kmt_test.h>
+
+#define NDEBUG
+#include <debug.h>
+
+/* GLOBALS ********************************************************************/
+
+/* Trigger a bug in the CORE-20078 */
+#ifdef _MSC_VER
+#pragma optimize("gst", on)
+#pragma auto_inline(on)
+#else
+#pragma GCC optimize("O3")
+#endif
+
+typedef struct _TEST_CONTEXT
+{
+    PVOID OldBuffer;
+    UCHAR Pad[78];
+    ULONG Port;
+    ULONG Size;
+    PVOID Buffer;
+} TEST_CONTEXT, *PTEST_CONTEXT;
+
+extern BOOLEAN
+VmIsVMware(VOID);
+
+/* FUNCTIONS ******************************************************************/
+
+static
+BOOLEAN
+VmIsNekoProject(VOID)
+{
+    UCHAR Input[4] = "NP2";
+    UCHAR Output[4] = { 0 };
+    UCHAR i;
+
+    for (i = 0; i < 3; i++)
+        WRITE_PORT_UCHAR((PUCHAR)0x7EF, Input[i]);
+
+    for (i = 0; i < 3; i++)
+        Output[i] = READ_PORT_UCHAR((PUCHAR)0x7EF);
+
+    return (*(PULONG)Input == *(PULONG)Output);
+}
+
+static
+BOOLEAN
+VmIsVbox(VOID)
+{
+    ULONG PciId, BytesRead;
+    PCI_SLOT_NUMBER Slot;
+
+    Slot.u.AsULONG = 0;
+    Slot.u.bits.DeviceNumber = 4;
+    Slot.u.bits.FunctionNumber = 0;
+
+    BytesRead = HalGetBusDataByOffset(PCIConfiguration,
+                                      0,
+                                      Slot.u.AsULONG,
+                                      &PciId,
+                                      0,
+                                      sizeof(PciId));
+    return (BytesRead == sizeof(PciId)) && (PciId == 0xCAFE80EE);
+}
+
+static
+BOOLEAN
+DetectVirtualMachine(VOID)
```

Comment:
`// TODO: QEMU/KVM, Virtual PC / Hyper-V` ?

Reply:
I've pushed a change that removes some of the code. Only VMs used by Testman are checked.